### PR TITLE
Fix concatenated key set validation

### PR DESCRIPTION
**Pull request by sbezverk**
Signed-off-by: Serguei Bezverkhi <sbezverk@cisco.com>

Current set key validation checks for a single nftMagic match, which breaks validation of concatenated keys. This PR adds new validation func and unit test.

### CODE REVIEW

**Comment by sbezverk** [comment]
@stapelberg Small fix in set validation, please take a look. Thank you.

File: set.go [hunk 1]
```diff
@@ -26,6 +26,10 @@ import (
 	"golang.org/x/sys/unix"
 )
 
+// SetConcateTypeBits defines concatination bits, originally defined in
+// https://git.netfilter.org/iptables/tree/iptables/nft.c#n999
```
Comment (stapelberg) [inline set.go]: Let’s make this a stable link: https://git.netfilter.org/iptables/tree/iptables/nft.c?id=26753888720d8e7eb422ae4311348347f5a05cb4#n1002

File: set.go [hunk 2]
```diff
@@ -26,6 +26,10 @@ import (
 	"golang.org/x/sys/unix"
 )
 
+// SetConcateTypeBits defines concatination bits, originally defined in
```
Comment (stapelberg) [inline set.go]: here and elsewhere: fix typo: SetConcatTypeBits (no “e” after Concat)

File: set.go [hunk 3]
```diff
@@ -464,6 +463,30 @@ func setsFromMsg(msg netlink.Message) (*Set, error) {
 	return &set, nil
 }
 
+func validateKeyType(bits uint32) ([]uint32, bool) {
+	var unpackTypes []uint32
+	var invalidTypes []uint32
+	found := false
+	valid := true
+	for bits != 0 {
+		unpackTypes = append(unpackTypes, bits&0x0000003f)
```
Comment (stapelberg) [inline set.go]: Why not just write 0x3f?

File: set_test.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+			nftMagicPacked: genSetKeyType(7, 13, 25),
+			pass:           false,
+			invalid:        []uint32{25},
+		},
+		{
+			name:           "Multiple nftMagic with 2 invalid",
+			nftMagicPacked: genSetKeyType(7, 13, 25, 26),
+			pass:           false,
+			invalid:        []uint32{26, 25},
+			// Invalid entries will apprear in reverse order
```
Comment (stapelberg) [inline set_test.go]: typo: apprear → appear

File: set.go (hunk 1 above)
Comment (sbezverk) [reply set.go]: done

File: set.go (hunk 2 above)
Comment (sbezverk) [reply set.go]: done

File: set.go (hunk 3 above)
Comment (sbezverk) [reply set.go]: done

File: set_test.go (hunk 4 above)
Comment (sbezverk) [reply set_test.go]: done

**Pull request merged**